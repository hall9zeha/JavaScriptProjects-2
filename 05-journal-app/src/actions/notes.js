import { db } from "../firebase/firebaseConfig";
import { types } from "../types/Types";

export const startNewNote = () =>{
    // getState es el segundo parámetro proveido por redux-thunk que nos permite despachar las acciones
    // y funciona muy parecido a useSelector al acceder a los objetos de nuestros estados en redux, para 
    // extraer en este caso los datos de usuario en sesión.
    return async (dispatch,getState)=>{
        const {uid} = getState().auth;

        const newNote = {
            title:'',
            body:'',
            date: new Date().getTime()
        }
        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote)
        // Un vez creada la nota en firebase solo necesitaremos su id y enviaremos nuevamente el objeto newNote
        // que está vacío hacia nuestra vista, para actualizar los cambios que hagamos y guardarlos
        dispatch(activeNote(docRef.id,newNote))
    }   
        
}

export const activeNote = (id, note) =>({
    type:types.notesActive,
    payload:{
        id,
        ...note
    }
})

export const setNotes = (notes) =>({
    type:types.notesLoad,
    payload:notes
})