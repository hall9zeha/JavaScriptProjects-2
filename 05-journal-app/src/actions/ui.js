import { types } from "../types/Types";

export const setError =  (error) =>({
    type:types.uiSetError,
    payload:error
});

export const removeError = () =>({
    type:types.uiRemoveError
})

export const uiStartLoading =()=>({
    type:types.uiStartLoading,
    payload:true
})
export const uiFinishLoading = ()=>({
    type:types.uifinishLoading,
    payload:false
})