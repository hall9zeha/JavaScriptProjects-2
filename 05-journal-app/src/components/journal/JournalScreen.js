import React from 'react'
import { Sidebar } from './Sidebar'

export const JournalScreen = () => {
  return (
    <div className='journal__main-content'> 
      <Sidebar/>
      <h1>JournalScreen</h1>
      
    </div>
  )
}
