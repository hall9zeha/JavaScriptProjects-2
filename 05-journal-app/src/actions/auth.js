// Las acciones no son más que simples funciones que representan nuestros casos de uso

import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/Types";
import { uiFinishLoading, uiStartLoading } from "./ui";



export const startLoginWithEmailPassword = (email, password) =>{
    return (dispatch)=>{
        dispatch(uiStartLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName));
                dispatch(uiFinishLoading())
            })
            .catch(e=>{
                dispatch(uiFinishLoading())
                console.log(e)
            })
        
    }
}

export const startLoginWithGoogle = ()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName))
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password,name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then(async({user}) =>{
                await user.updateProfile({displayName:name});
                dispatch(login(user.uid,user.displayName));
            })
            .catch(e=>{
                console.log(e)
            })
    }
}


// La siguiente función retornará un objeto
export const login = (uid, displayName)=>({
    type: types.login,
    payload:{
        uid,
        displayName
    }
})

// Logout

export const startLogout = ()=>{
    return async (dispatch)=>{
        await firebase.auth().signOut();
        dispatch(logout())
    }
}

export const logout=()=>({
    type:types.logout
})