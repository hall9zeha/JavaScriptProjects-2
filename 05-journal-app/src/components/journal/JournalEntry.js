import moment from 'moment'
import React from 'react'
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = (note) => {
  const {id,date,title,body,url} = note;
  const customDate = moment(date);
  const dispatch = useDispatch();

  const handleEntryClick=()=>{
    dispatch(
      activeNote(id,note)
    )
  }

  return (
    <div 
      onClick={handleEntryClick}
      className='journal__entry pointer animate__animated animate__fadeIn animated__faster'>
      {
        // Si la url existe y no es undefined o null mostrar 
        url &&
        <div
          className='journal__entry-picture'
          style={{
            backgroundSize:'cover',
            backgroundImage:`url(${url})`
          }}
        ></div>
      }

      <div className='journal__entry-body'>
        <p className='journal__entry-title'>{title}</p>
        <p className='journal__entry-content'>
          {body}
          </p>
      </div>
      <div className='journal__entry-date-box'>
          <span>{customDate.format('dddd')}</span>
          <h4>{customDate.format('Do')}</h4>
      </div>
    </div>
  )
}
