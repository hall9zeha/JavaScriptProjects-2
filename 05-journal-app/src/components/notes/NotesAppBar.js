import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {
  const dispatch = useDispatch();
  const {active:currentNote} = useSelector(state=>state.notes)

  const handleSave=()=>{
    dispatch(startSaveNote(currentNote))
  }

  const handlePictureUpload = ()=>{
    document.querySelector('#fileSelector').click()
  }
  const handleFileChange = (e)=>{
    
    const file = e.target.files[0]
    if(file){
      dispatch(startUploading(file))
    }
  }
  
  
  return (
    <div className='notes__appbar'>
        <span>13 julio 2024</span>
        <input
          id='fileSelector'
          type='file'
          name='file'
          style={{display:'none'}}
          onChange={handleFileChange}
        />
        <div>
            <button 
              onClick={handlePictureUpload}
              className='btn'>Picture</button>
            <button 
              onClick={handleSave}
              className='btn'>Save</button>
        </div>
    </div>
  )
}
