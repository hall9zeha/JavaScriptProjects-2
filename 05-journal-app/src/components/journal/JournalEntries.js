import React from 'react'
import { JournalEntry } from './JournalEntry'
import { useSelector } from 'react-redux'

export const JournalEntries = () => {
  /*
  La ventaja de usar redux es que en la mayoría de los casos no necesitamos enviar argumentos a través de las properties
  solamente debemos acceder a nuestros estados para obtener la información de allí
  */
  const {notes} =useSelector(state=>state.notes);

  return (
    <div className='journal__entries'>
        {
            notes.map(note=>(
                <JournalEntry key={note.id}
                  //Enviamos cada nota como argumento
                  {...note}
                />
            ))
        }
    </div>
  )
}
