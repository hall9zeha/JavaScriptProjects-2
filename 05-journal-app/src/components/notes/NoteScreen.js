import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { activeNote } from '../../actions/notes'

export const NoteScreen = () => {

    const dispatch = useDispatch();
    //Renombramos el objeto desestructurado 'active' a 'note' active:note
    const {active:note} = useSelector(state =>state.notes)
    const [formValues, handleInputChange, reset] = useForm(note);
    const {body, title} = formValues

    /*
    Para que los campos del formulario de nuestra nota cambien al seleccionar otra
    debemos resetear el formulario con los nuevos valores, para ello usaremos lo siguiente
    
    useRef almacena una variable mutable y no va aredibujar el componente si esta cambia
    */
    const activeNoteId =  useRef(note.id);
    useEffect(()=>{
        //Si el id de la nota seleccionada es diference a nuestro id inicial guardado
        //entonces reetear el formulario y establecer activeNoteId con el nuevo valor
        if(note.id !== activeNoteId.current){
            reset(note);
            activeNoteId.current = note.id
        } 
    },[note,reset])

    //Otro useEffect para actualizar el estado de la nota activa cuando escribimos algo dentro de sus campos
    useEffect(()=>{
        dispatch(activeNote(formValues.id,{...formValues}))
    },[formValues,dispatch])

  return (
    <div className='notes__main-content'>
        <NotesAppBar/>
        <div className='notes__content'>
            <input 
                type='text'
                placeholder='Some awesome'
                className='notes__title-input'
                name='title'
                value={title}
                onChange={handleInputChange}
                />
            <textarea
                placeholder='What happened today'
                className='notes__textarea'
                name='body'
                value={body}
                onChange={handleInputChange} 
            
            ></textarea>
            {
                note.url &&
                    <div className='notes__image'>
                        <img
                        src={note.url}
                        alt='nature road'
                        />
                    </div>
            }
           
        </div>
    </div>
  )
}
