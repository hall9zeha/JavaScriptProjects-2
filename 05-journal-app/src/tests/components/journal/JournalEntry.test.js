/**
 * @jest-environment jsdom
 */
import {Provider} from 'react-redux';
import renderer ,{act}from 'react-test-renderer';


import configureStore from 'redux-mock-store'
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

const thunk = require('redux-thunk').thunk;


//***************************/
const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
    auth:{
        id:'123',
        name:'Martha'
    },
    ui:{
        loading:false,
        msgError:null
    },
    notes:{
        active:{id:'Abs'},
        notes:[]
    }
}
let store = mockStore(initState)
store.dispatch=jest.fn()

const note = {
    id:24,
    date:2024,
    title:'Hello',
    body:'world',
    url:'https://mockweb.com/image.jpg'
}

describe('Test in <Journalentry/> component', () => { 
    const wrapper = renderer.create(
        <Provider store={store}>
            <JournalEntry {...note}/>
        </Provider>
    )
    test('should show successfully', () => { 
        expect(wrapper).toMatchSnapshot()
     })
    test('should activate note', () => {
        const newNote  = wrapper.root.findByProps({className:'journal__entry pointer animate__animated animate__fadeIn animated__faster'});
        newNote.props.onClick();
        //Con activeNote evaluamos que se haya activado la misma nota que hemos creado en la parte superior
        //Ya que nos devuelve una acci√≥n como la siguiente:
        /*
            {
                type: '[Notes] Set active note',
                payload: {
                    id: 24,
                    date: 2024,
                    title: 'Hello',
                    body: 'world',
                    url: 'https://mockweb.com/image.jpg'
                    }
            }
        
        */
        expect(store.dispatch).toHaveBeenCalledWith(activeNote(note.id,{...note}))
    
    })
})