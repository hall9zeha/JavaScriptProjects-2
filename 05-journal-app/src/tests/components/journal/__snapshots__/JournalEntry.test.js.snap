// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Journalentry/> component should show successfully 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animated__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://mockweb.com/image.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hello
    </p>
    <p
      className="journal__entry-content"
    >
      world
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
